docs/input-validation.rst

```diff
@@ -0,0 +1,129 @@
+
+#skipped (validator unavailable)
+
+    a backend command for a validator command is not available.
+
+*validator* file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*validator* file in a *Units* test directory specifies which
+validator the make target should use.
+
+.. code-block:: console
+
+  $ cat Units/simple-json.d/validator
+  jq
+
+If you put *validator* file to a category directory (a directory
+having *.r* suffix), the make target uses the validator specified in
+the file as default.  The default validator can be overridden with a
+*validator* file in a subdirectory.
+
+.. code-block:: console
+
+  $ cat Units/parser-puppetManifest.r/validator
+  puppet
+  # cat Units/parser-puppetManifest.r/puppet-append.d/validator
+  KNOWN-INVALIDATION
+
+In the example, the make target uses *puppet* validator for validating
+the most of all input files under *Units/parser-puppetManifest.r*
+directory. An exception is an input file under
+*Units/parser-puppetManifest.r/puppet-append.d* directory.  The
+directory has its specific *validator* file.
+
+If a *Unit* test case doesn't have *expected.tags* file, the make
+target doesn't run the validator on the file even if a default
+validator is given in its category directory.
+
+If a *Unit* test case specifies KNOWN-INVALIDATION in its *validator*
+file, the make target just increments "#skipped (known invalidation)"
+counter. The target reports the counter at the end of execution.
+
+validator command
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A validator specified in a *validator* file is a command file put
+under *misc/validators* directory.  The command must have "validator-"
+as prefix in its file name. For an example,
+*misc/validators/validator-jq* is the command for "jq".
+
+The command file must be an executable. *validate-input* make target
+run the command in two ways.
+
+*is_runnable* method
+
+    Before running the command as a validator, the target runs
+    the command with "is_runnable" as the first argument.
+    A validator command can let the target know whether the
+    validator command is runnable or not with exit status.
+    0 means ready to run. Non-zero means not ready to run.
+
+    The make target never runs the validator command for
+    validation purpose if the exit status is non-zero.
+
+    For an example, *misc/validators/validator-jq* command uses *jq*
+    command as its backend. If *jq* command is not available on a
+    system, *validator-jq* can do nothing.  If such case,
+    *is_runnable* method of *validator-jq* command should exit with
+    non-zero value.
+
+*validate* method
+
+    The make targets runs the command with "validate* and an input
+    file name for validating the input file.  The command exits
+    non-zero if the input file is written invalid syntax. It exits
+    non-zero for invalid input.  This method will never run if
```

Comment:
Instead of : 
```The command exits
    non-zero if the input file is written invalid syntax. It exits
    non-zero for invalid input.
 ```

The command exits non-zero if the input file contains invalid syntax.

I do not think you need to repeat yourself.